crates/ty_python_semantic/resources/mdtest/type_qualifiers/final.md

```diff
@@ -231,7 +227,86 @@ FINAL_LIST: Final[list[int]] = [1, 2, 3]
 FINAL_LIST[0] = 4
 ```
 
-## Too many arguments
+## Overriding in subclasses
+
+When a symbol is qualified with `Final` in a class, it cannot be overridden in subclasses.
+
+```py
+from typing import Final
+
+class Base:
+    FINAL_A: Final[int] = 1
+    FINAL_B: Final[int] = 1
+    FINAL_C: Final = 1
+
+class Derived(Base):
+    # TODO: This should be an error
+    FINAL_A = 2
+    # TODO: This should be an error
+    FINAL_B: Final[int] = 2
+    # TODO: This should be an error
+    FINAL_C = 2
+```
+
+## Syntax and usage
+
+### Legal syntactical positions
+
+Final may only be used in assignments or variable annotations. Using it in any other position is an
+error.
+
+```py
+from typing import Final, ClassVar, Annotated
+
+LEGAL_A: Final[int] = 1
+LEGAL_B: Final = 1
+LEGAL_C: Final[int]
+LEGAL_C = 1
+LEGAL_D: Final
+LEGAL_D = 1
+
+class C:
+    LEGAL_E: ClassVar[Final[int]] = 1
+    LEGAL_F: Final[ClassVar[int]] = 1
+    LEGAL_G: Annotated[Final[ClassVar[int]], "metadata"] = 1
+
+    def __init__(self):
+        self.LEGAL_H: Final[int] = 1
+        self.LEGAL_I: Final[int]
+        self.LEGAL_I = 1
+
+# TODO: This should be an error
+def f(ILLEGAL: Final[int]) -> None:
+    pass
+```
+
+### Attribute assignment outside `__init__`
+
+Qualifying an instance attribute with `Final` outside of `__init__` is not allowed. The instance
+attribute must be assigned only once, when the instance is created.
```

Comment:
how does `__new__` play into things here? (I can't remember what the spec says.) E.g. this feels ~fine to me... but if the spec disallows it, then the spec disallows it, I guess:

```py
from typing import Final

class Foo:
    def __new__(cls):
        self = object.__new__(cls)
        self.x: Final = 42
        return self
```

it's basically the pattern that `fractions.Fraction` uses in the stdlib: https://github.com/python/cpython/blob/9a21df7c0a494e2819775eabd522ebec994d96c0/Lib/fractions.py#L205-L311

Reply:
The spec says (emphasis mine):

> Finally, as `self.id: Final = 1` (also optionally with a type in square brackets). **This is allowed only in `__init__` methods**, so that the final instance attribute is assigned only once when an instance is created.

In your code snippet in particular, any declaration would be illegal in that position. You can't annotate attribute assignments if they don't refer to the first parameter in a method (ty does not yet enforce this, see https://github.com/astral-sh/ty/issues/509).